utils.go

```diff
@@ -82,25 +83,39 @@ func GetRootlessRuntimeDir(rootlessUID int) (string, error) {
 	return path, nil
 }
 
+func isRootlessRuntimeDirOwner(dir string) bool {
+	st, err := system.Lstat(dir)
+	return err == nil && int(st.UID()) == os.Getuid() && st.Mode()&0700 == 0700 && st.Mode()&0066 == 0000
+}
+
 func getRootlessRuntimeDir(rootlessUID int) (string, error) {
 	runtimeDir, err := homedir.GetRuntimeDir()
 	if err == nil {
 		return runtimeDir, nil
 	}
-	tmpDir := fmt.Sprintf("/run/user/%d", rootlessUID)
-	st, err := system.Stat(tmpDir)
-	if err == nil && int(st.UID()) == os.Getuid() && st.Mode()&0700 == 0700 && st.Mode()&0066 == 0000 {
-		return tmpDir, nil
+	initCommand, err := ioutil.ReadFile("/proc/1/comm")
+	if err != nil || string(initCommand) == "systemd" {
+		tmpDir := fmt.Sprintf("/run/user/%d", rootlessUID)
+		if isRootlessRuntimeDirOwner(tmpDir) {
+			return tmpDir, nil
+		}
 	}
-	tmpDir = fmt.Sprintf("%s/%d", os.TempDir(), rootlessUID)
-	if err := os.MkdirAll(tmpDir, 0700); err != nil {
-		logrus.Errorf("failed to create %s: %v", tmpDir, err)
-	} else {
+	tmpDir := fmt.Sprintf("%s/containers-user-%d", os.TempDir(), rootlessUID)
+	if _, err := os.Stat(tmpDir); os.IsNotExist(err) {
+		newTmpDir, err := ioutil.TempDir(os.TempDir(), "containers-user-*")
+		if err != nil {
+			logrus.Errorf("failed to create temp directory: %v", err)
+		}
+		if err := os.Rename(newTmpDir, tmpDir); err != nil {
+			logrus.Errorf("failed to rename %s to %s: %v", newTmpDir, tmpDir, err)
```

Comment:
I don’t see what the `TempDir` + `Rename` does that straight `Mkdir` doesn’t.

Reply:
I don't know too. Please write test in go that clearly defines what problem we need to handle. Because without it we can work on this pull request forever.

Reply:
“I don’t know why the code is here” is a potential security risk for my review.

Reply:
I've read previous comments about security problem and implemented solution that everyone agree about. Now another security problem appeared (as I see) so I am requesting people to **describe security problem in actual code**.

Reply:
I've read it about 10 times, I see no position anywhere which can be kept or changed. You points to some attack, but nobody can declear actual attack.

Reply:
At this point I’m not pointing at a specific attack; I think that a plain `os.Mkdir` (not `All`!) with a strict error check would work fine.

If Dan is asking for a `TmpDir`+`Rename`, I don’t know why that is necessary, which indicates I may well be making a mistake, and missing another attack I should be thinking about.

Reply:
I see the only possible attack - user attacks himself. We can't do anything with such user - he can attack himself in any folder with sticky bit or without it.

Reply:
Well, either this approach is necessary, in which case I will need to rethink things, or it is not, and the code should be simpler.

Let’s wait for Dan.

Reply:
Yes mkdir should do the right thing.  IE Create the directory with the correct permissions, and fail if an attacker has created the directory since you checked if it exists.  The rename is overkill, sorry for suggesting it.